### DISCUSSION TITLE
Strange gradient behavior

### DISCUSSION THREAD

**Discussion by benjamin-reilly** [idaholab/moose#28660] [Q&A Modules: Phase field]
Hello,

New user here, so I apologize in advance if this issue is a simple user error.

I am experiencing what I ***think*** is weird behavior when I have MOOSE compute the gradient of a variable. In short, I am interested in multi-phase-field modeling of non-conserved order parameters, using in particular the model outlined [here](https://www.sciencedirect.com/science/article/abs/pii/S0167278999001293) so as to maintain the ability to specify different mobilities for different interfaces. This model depends on using indicator functions encoding the ***local*** number of phases with non-zero gradients, and effectively rescales the mobility in accordance with this local information. Moreover, in pure-phase regions the mobility can effectively be viewed as being set to 0 (of course, you could view this as being the case in most non-conserved phase-field models, but it is especially important in this model when generalizing to > 2 phases, and I am pointing it out here so that my example input file included below makes a bit more sense).

So, I am trying to use the norms of the gradients of my variables in order to indicate where these non-zero gradients exist (defining the indicator function simply as ```if(norm>0,1,0)```). However, my gradient norms do not appear to be accurately representing the actual gradient norms of my variables. To simplify the explanation, I am including screenshots below with brief descriptions:

First, the initial state of one of my phase fields:
![eta0_t0](https://github.com/user-attachments/assets/0213f1d1-3a4f-4608-b0da-6990017c2494)

The corresponding L2 norm of the gradient of this phase field:
![norm0_t0](https://github.com/user-attachments/assets/bbf13bac-091d-48b8-9e8e-370e1ba0395c)

Some time later, the phase field has not changed too much (although it is slightly squeezed, since it is hitting a region of the domain in which the mobility is 0-- even though the mobility ***should*** be updated to a non-zero value when the gradient becomes non-zero as the interface approaches this region):
![eta0_t1](https://github.com/user-attachments/assets/fc329742-d1af-4993-82d9-4f6cd5814ac7)

The corresponding gradient norm at this same time, which has decreased relative to the previous time (even though, if anything, it should have ***increased*** due to the interface being slightly more confined):
![norm0_t1](https://github.com/user-attachments/assets/2e102d4e-7090-45df-aac9-7b543bebab70)

Aside from it being concerning that the norms calculated by MOOSE do not appear to accurately be the gradient norms of the variables, this is especially problematic for my model because I would like to update the mobility to be non-zero as the interface advances into new regions, which is not occurring at all. Instead, the only non-zero mobilities throughout the simulation are located at the initial location of the interface, i.e. the "update" via investigating the new gradient norms is never actually occurring (and I am guessing that this is most likely related to the gradient norms misbehaving).

I am in particular using the VariableGradientMaterial material in order to compute the gradient norms, but I have experimented with other methods as well (e.g. using the VariableGradientComponent AuxKernel and then manually constructing the gradient norms in ParsedMaterial materials), to no avail. So, I think there is something I probably just don't understand regarding how MOOSE is computing the gradients of variables.

Below I am including an input file that I have created as a minimal reproducible example of this behavior. Note that I am ultimately concerned with simulating multi-phase-field models, and so even though the below input file is effectively only a single-order-parameter model, I am still using e.g. ACMultiInterface in order to incorporate the gradient energy term (so as to make this example consistent with the more complicated systems I am actually interested in).

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 128
  ny = 16
  xmin = -120
  xmax = 120
  ymin = 0
  ymax = 30
  elem_type = QUAD
[]

[GlobalParams]
outputs = exodus
[]

[ICs]
  [eta0_IC]
    type = FunctionIC
    variable = eta0
    function = eta0_IC
  []
  [eta1_IC]
    type = FunctionIC
    variable = eta1
    function = eta1_IC
  []
[]

[AuxVariables]
[]

[AuxKernels]
[]

[Functions]
  [eta0_IC]
    type = ParsedFunction
    expression = 'if(x > -width/2 & x < width/2, 0.5 * (1 + sin(3.14159265359*x/width)), if(x < -width/2,0,1))'
    symbol_names = 'width'
    symbol_values = '20'
  [../]
  [eta1_IC]
    type = ParsedFunction
    expression = '1-eta0_IC'
    symbol_names = ' eta0_IC'
    symbol_values = ' eta0_IC'
  [../]
[]

[Variables]
  [./eta0]
    order = SECOND
    family = MONOMIAL
  [../]
  [./eta1]
    order = SECOND
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./deta0dt]
    type = TimeDerivative
    variable = eta0
  [../]
  [./ACBulk0_self]
    type = AllenCahn
    variable = eta0
    coupled_variables = ' eta0 eta1'
    mob_name = L0
    f_name = F0
  [../]
  [./ACInterface0_self]
    type = ACMultiInterface
    variable = eta0
    etas = ' eta0 eta1'
    mob_name = L0
    kappa_names = ' kappa00 kappa01'
  [../]
  [./simplexKernel]
    type = MaterialPropertyValue
    prop_name = SimplexMaterial
    variable = eta1
  [../]
[]

[BCs]
  [NeumannBC0]
    type = MatNeumannBC
    variable = eta0
    boundary = 'top bottom left right'
    boundary_material = norm0
    value = 0.
  []
  [NeumannBC1]
    type = MatNeumannBC
    variable = eta1
    boundary = 'top bottom left right'
    boundary_material = norm1
    value = 0.
  []
[]

[Materials]
  [./consts]
    type = GenericConstantMaterial
    prop_names = ' kappa00 kappa01 kappa10 kappa11'
    prop_values = ' 5.836100177798656 5.836100177798656 5.836100177798656 5.836100177798656'
  [../]
  [./mobility_base]
    type = GenericConstantMaterial
    prop_names  = 'L_base'
    prop_values  = '0.25'
  [../]
  [./mobility0]
    type = ParsedMaterial
    property_name  = L0
    expression = 'if(norm0>0&norm1>0,L_base,0.)'
    material_property_names = 'L_base  norm0(eta0) norm1(eta1)'
  [../]
  [./simplexMaterial]
    type = DerivativeParsedMaterial
    property_name = SimplexMaterial
    coupled_variables = ' eta0'
    expression = '1.-eta0'
  [../]
  [./etasum]
    type = ParsedMaterial
    property_name = etasum
    coupled_variables = ' eta0 eta1'
    expression = ' eta0 + eta1'
  [../]
  [./phase_free_energy0]
    type = DerivativeParsedMaterial
    property_name = F0
    expression = 'interp:=if(eta0>0&eta0<1,(asin(2*eta0-1)+(4*eta0-2)*sqrt(eta0*(1-eta0)))/3.14159265359,0.); 0.07200000000000001*eta0*eta1*if(eta0*eta1<0,-1.,1.) + -3.58918109*interp'
    coupled_variables = ' eta0 eta1'
    constant_names = 'T_m_c'
    constant_expressions = ' 1687'
  [../]
  [./norm0]
    type = VariableGradientMaterial
    prop = norm0
    variable = eta0
  [../]
  [./norm1]
    type = VariableGradientMaterial
    prop = norm1
    variable = eta1
  [../]
[]

[Postprocessors]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'implicit-euler'
  solve_type = 'NEWTON'
  automatic_scaling = true
  dtmin = 1e-10
  l_max_its = 50
  l_tol = 1.0e-1
  nl_max_its = 10
  nl_rel_tol = 1.0e-1
  start_time = 0
  dt = 1
  num_steps = 15
  dtmax = 1
[]

[Debug]
[]

[Outputs]
  time_step_interval = 1
  [exodus]
    type = Exodus
  []
  [./table]
    type = CSV
    delimiter = ' '
  [../]
[]
```

**Comment by GiudGiud** [comment]
Hello

I dont see any auxvariable or auxkernel. How are you obtaining the norm of the gradient? With paraview postprocessing?

**Reply by benjamin-reilly** [reply]
Hi Guillaume,

Thanks for the quick reply. I am visualizing the norm of the gradient from the output exodus file in Peacock.

For the sake of completeness, here's the alternative version I alluded to in my previous message (using the VariableGradientComponent AuxKernel with AuxVariables to store the gradient components), which suffers from the same issues:

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 128
  ny = 16
  xmin = -120
  xmax = 120
  ymin = 0
  ymax = 30
  elem_type = QUAD
[]

[GlobalParams]
outputs = exodus
[]

[ICs]
  [eta0_IC]
    type = FunctionIC
    variable = eta0
    function = eta0_IC
  []
  [eta1_IC]
    type = FunctionIC
    variable = eta1
    function = eta1_IC
  []
[]

[AuxVariables]
  [./eta0_gradx]
    order = FIRST
    family = MONOMIAL
  [../]
  [./eta0_grady]
    order = FIRST
    family = MONOMIAL
  [../]
  [./eta1_gradx]
    order = FIRST
    family = MONOMIAL
  [../]
  [./eta1_grady]
    order = FIRST
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./eta0_gradx_kernel]
    type = VariableGradientComponent
    gradient_variable = eta0
    variable = eta0_gradx
    component = x
    execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
  [../]
  [./eta0_grady_kernel]
    type = VariableGradientComponent
    gradient_variable = eta0
    variable = eta0_grady
    component = y
    execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
  [../]
  [./eta1_gradx_kernel]
    type = VariableGradientComponent
    gradient_variable = eta1
    variable = eta1_gradx
    component = x
    execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
  [../]
  [./eta1_grady_kernel]
    type = VariableGradientComponent
    gradient_variable = eta1
    variable = eta1_grady
    component = y
    execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
  [../]
[]

[Functions]
  [eta0_IC]
    type = ParsedFunction
    expression = 'if(x > -width/2 & x < width/2, 0.5 * (1 + sin(3.14159265359*x/width)), if(x < -width/2,0,1))'
    symbol_names = 'width'
    symbol_values = '20'
  [../]
  [eta1_IC]
    type = ParsedFunction
    expression = '1-eta0_IC'
    symbol_names = ' eta0_IC'
    symbol_values = ' eta0_IC'
  [../]
[]

[Variables]
  [./eta0]
    order = SECOND
    family = MONOMIAL
  [../]
  [./eta1]
    order = SECOND
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./deta0dt]
    type = TimeDerivative
    variable = eta0
  [../]
  [./ACBulk0_self]
    type = AllenCahn
    variable = eta0
    coupled_variables = ' eta0 eta1'
    mob_name = L0
    f_name = F0
  [../]
  [./ACInterface0_self]
    type = ACMultiInterface
    variable = eta0
    etas = ' eta0 eta1'
    mob_name = L0
    kappa_names = ' kappa00 kappa01'
  [../]
  [./simplexKernel]
    type = MaterialPropertyValue
    prop_name = SimplexMaterial
    variable = eta1
  [../]
[]

[BCs]
  [NeumannBC0]
    type = MatNeumannBC
    variable = eta0
    boundary = 'top bottom left right'
    boundary_material = norm0
    value = 0.
  []
  [NeumannBC1]
    type = MatNeumannBC
    variable = eta1
    boundary = 'top bottom left right'
    boundary_material = norm1
    value = 0.
  []
[]

[Materials]
  [./consts]
    type = GenericConstantMaterial
    prop_names = ' kappa00 kappa01 kappa10 kappa11'
    prop_values = ' 5.836100177798656 5.836100177798656 5.836100177798656 5.836100177798656'
  [../]
  [./mobility_base]
    type = GenericConstantMaterial
    prop_names  = 'L_base'
    prop_values  = '0.25'
  [../]
  [./mobility0]
    type = ParsedMaterial
    property_name  = L0
    expression = 'if(norm0>0&norm1>0,L_base,0.)'
    material_property_names = 'L_base  norm0(eta0_gradx,eta0_grady) norm1(eta1_gradx,eta0_grady)'
  [../]
  [./simplexMaterial]
    type = DerivativeParsedMaterial
    property_name = SimplexMaterial
    coupled_variables = ' eta0'
    expression = '1.-eta0'
  [../]
  [./etasum]
    type = ParsedMaterial
    property_name = etasum
    coupled_variables = ' eta0 eta1'
    expression = ' eta0 + eta1'
  [../]
  [./phase_free_energy0]
    type = DerivativeParsedMaterial
    property_name = F0
    expression = 'interp:=if(eta0>0&eta0<1,(asin(2*eta0-1)+(4*eta0-2)*sqrt(eta0*(1-eta0)))/3.14159265359,0.); 0.07200000000000001*eta0*eta1*if(eta0*eta1<0,-1.,1.) + -3.58918109*interp'
    coupled_variables = ' eta0 eta1'
    constant_names = 'T_m_c'
    constant_expressions = ' 1687'
  [../]
  [./norm0]
    type = DerivativeParsedMaterial
    property_name = norm0
    coupled_variables = 'eta0_gradx eta0_grady'
    expression = 'sqrt(eta0_gradx^2 + eta0_grady^2)'
  [../]
  [./norm1]
    type = DerivativeParsedMaterial
    property_name = norm1
    coupled_variables = 'eta1_gradx eta1_grady'
    expression = 'sqrt(eta1_gradx^2 + eta1_grady^2)'
  [../]
[]

[Postprocessors]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'implicit-euler'
  solve_type = 'NEWTON'
  automatic_scaling = true
  dtmin = 1e-10
  l_max_its = 50
  l_tol = 1.0e-1
  nl_max_its = 10
  nl_rel_tol = 1.0e-1
  #nl_abs_tol = 1.0e-3
  start_time = 0
  dt = 1
  num_steps = 15
  dtmax = 1
[]

[Debug]
[]

[Outputs]
  time_step_interval = 1
  [exodus]
    type = Exodus
  []
  [./table]
    type = CSV
    delimiter = ' '
  [../]
[]
```

**Reply by GiudGiud** [reply]
so the VariableGraidentAux compute the gradient within each element

but you are using monomials, which are discontinuous. So the gradient on the faces is infinite and accounts for a good chunk of the jump in values most likely

**Reply by GiudGiud** [reply]
You could try this to confirm. it will have a projection error but it might bring some insights

- project eta onto a continuous variable (eta-nodal with 2nd order lagrange for example)
- use the VariableGradientComponentAux on eta-nodal

And check:
- if the gradient does appear more pronounced for the second case than the first one.
- if the integral of the gradient does retrieve the expected jump in values eta_right - eta_left

**Reply by benjamin-reilly** [reply]
I understand your suggestion, however I think I currently lack the MOOSE experience to be able to effectively implement such a projection.

Nevertheless, your suggestion reminded me of your response [here](https://github.com/idaholab/moose/discussions/26693), which I had come across before making this post. Sure enough, after I switched the gradient component AuxVariables to be constant monomials, the gradients started looking accurate! I'm not sure why this didn't work the first time I came across your response in that other thread-- I was probably tweaking too many things simultaneously in the input file to properly isolate the issue.

In any case, things are looking much better now. Thanks a bunch for your help!
